prow/config/secret/agent.go

```diff
@@ -98,81 +97,59 @@ func Censor(content []byte) []byte {
 // agent watches a path and automatically loads the secrets stored.
 type agent struct {
 	sync.RWMutex
-	secretsMap map[string][]byte
+	secretsMap map[string]secretReloader
 	*secretutil.ReloadingCensorer
 }
 
+type secretReloader interface {
+	getRaw() []byte
+	start(reloadCensor func()) error
+}
+
 // Add registers a new path to the agent.
 func (a *agent) Add(path string) error {
-	secret, err := loadSingleSecret(path)
-	if err != nil {
+	return a.add(path, &parsingSecretReloader[[]byte]{
+		path:      path,
+		parsingFN: func(b []byte) ([]byte, error) { return b, nil },
+	})
+}
+
+func (a *agent) add(path string, loader secretReloader) error {
+	if err := loader.start(a.refreshCensorer); err != nil {
 		return err
 	}
 
-	a.setSecret(path, secret)
+	a.setSecret(path, loader)
 
-	// Start one goroutine for each file to monitor and update the secret's values.
-	go a.reloadSecret(path)
 	return nil
 }
 
-// reloadSecret will begin polling the secret file at the path. If the first load
-// fails, Start with return the error and abort. Future load failures will log
-// the failure message but continue attempting to load.
-func (a *agent) reloadSecret(secretPath string) {
-	var lastModTime time.Time
-	logger := logrus.NewEntry(logrus.StandardLogger())
-
-	skips := 0
-	for range time.Tick(1 * time.Second) {
-		if skips < 600 {
-			// Check if the file changed to see if it needs to be re-read.
-			secretStat, err := os.Stat(secretPath)
-			if err != nil {
-				logger.WithField("secret-path", secretPath).
-					WithError(err).Error("Error loading secret file.")
-				continue
-			}
-
-			recentModTime := secretStat.ModTime()
-			if !recentModTime.After(lastModTime) {
-				skips++
-				continue // file hasn't been modified
-			}
-			lastModTime = recentModTime
-		}
-
-		if secretValue, err := loadSingleSecret(secretPath); err != nil {
-			logger.WithField("secret-path: ", secretPath).
-				WithError(err).Error("Error loading secret.")
-		} else {
-			a.setSecret(secretPath, secretValue)
-			skips = 0
-		}
-	}
-}
-
 // GetSecret returns the value of a secret stored in a map.
 func (a *agent) GetSecret(secretPath string) []byte {
 	a.RLock()
 	defer a.RUnlock()
-	return a.secretsMap[secretPath]
+	if val, set := a.secretsMap[secretPath]; set {
+		return val.getRaw()
+	}
+	return nil
 }
 
 // setSecret sets a value in a map of secrets.
-func (a *agent) setSecret(secretPath string, secretValue []byte) {
+func (a *agent) setSecret(secretPath string, secretValue secretReloader) {
 	a.Lock()
-	defer a.Unlock()
 	a.secretsMap[secretPath] = secretValue
+	a.Unlock()
 	a.refreshCensorer()
 }
 
 // refreshCensorer should be called when the lock is held and the secrets map changes
```

Comment:
this comment is not up-to-date any more, the lock is not held when this function is called above